workflow/controller/operator_test.go

```diff
@@ -10335,3 +10335,91 @@ func TestWorkflowNeedReconcile(t *testing.T) {
 		assert.Equal(t, "steps-need-reconcile", pods.Items[1].Spec.Containers[1].Env[0].Value)
 	}
 }
+
+var wfHasContainerSet = `
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  name: lovely-rhino
+spec:
+  templates:
+    - name: init
+      dag:
+        tasks:
+          - name: A
+            template: run
+            arguments: {}
+    - name: run
+      containerSet:
+        containers:
+          - name: main
+            image: alpine:latest
+            command:
+              - /bin/sh
+            args:
+              - '-c'
+              - sleep 9000
+            resources: {}
+          - name: main2
+            image: alpine:latest
+            command:
+              - /bin/sh
+            args:
+              - '-c'
+              - sleep 9000
+            resources: {}
+  entrypoint: init
+  arguments: {}
+  ttlStrategy:
+    secondsAfterCompletion: 300
+  podGC:
+    strategy: OnPodCompletion`
+
+// TestContainerSetDoesNotStopContainerWhenPodRemoved test whether a workflow has ContainerSet error when pod removed.
+func TestContainerSetDoesNotStopContainerWhenPodRemoved(t *testing.T) {
+	cancel, controller := newController()
+	defer cancel()
+	ctx := context.Background()
+	wfcset := controller.wfclientset.ArgoprojV1alpha1().Workflows("")
+	wf := wfv1.MustUnmarshalWorkflow(wfHasContainerSet)
+	wf, err := wfcset.Create(ctx, wf, metav1.CreateOptions{})
+	assert.Nil(t, err)
+	wf, err = wfcset.Get(ctx, wf.ObjectMeta.Name, metav1.GetOptions{})
+	assert.Nil(t, err)
+	woc := newWorkflowOperationCtx(wf, controller)
+	woc.operate(ctx)
+	pods, err := listPods(woc)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(pods.Items))
+
+	// mark pod Running
+	makePodsPhase(ctx, woc, apiv1.PodRunning)
+	woc = newWorkflowOperationCtx(woc.wf, controller)
+	woc.operate(ctx)
+	for _, node := range woc.wf.Status.Nodes {
+		if node.Type == wfv1.NodeTypePod {
+			assert.Equal(t, wfv1.NodeRunning, node.Phase)
+		}
+	}
+
+	// delete pod
+	time.Sleep(10 * time.Second)
```

Comment:
use
`_ = os.Setenv("RECENTLY_STARTED_POD_DURATION", "0")`
and 
set graceperiod to 0 on `metav1.DeleteOptions{` in `deletePods` function

so that we can get rid of time.sleep here

Reply:
Yeah,  I add it. Thanks!

Reply:
> use
> `_ = os.Setenv("RECENTLY_STARTED_POD_DURATION", "0")`

This sets it for the whole process though, no? not just this single test?

Reply:
Sigle and i remove it after test.

Reply:
That's not quite sufficient. Env vars are set for an entire process -- meaning it affects any tests that are ran in parallel and use the same env var. But I wasn't sure if Go had any special handling for this given that I've seen this in other tests. According to Go's own docs (for the [`t.Setenv` function](https://pkg.go.dev/testing#B.Setenv)), it does not have any special handling and this would indeed affect any parallel tests.

While this is done in other tests in this codebase ([50 occurrences](https://github.com/search?q=repo%3Aargoproj%2Fargo-workflows%20os.setenv&type=code)) which probably need a larger refactoring, this env var is a bit more global in its effects.

This is one of those reasons why config shouldn't be set purely in global-scoped env vars, as it can lead to races like this, particularly with test code that relies on globals.

Reply:
removed it, and i test sleep 5s is not enough. need 10s

Reply:
hmm, not needing to `sleep` in the tests is good to have though... @tczhao what do you think? 

perhaps we leave one in and then refactor in a separate PR to avoid globals?

Reply:
> removed it, and i test sleep 5s is not enough. need 10s

The 10s is due to `RECENTLY_STARTED_POD_DURATION` defaults to 10s,
`podReconciliation` will not update the wf status until `RECENTLY_STARTED_POD_DURATION` elapsed
https://github.com/argoproj/argo-workflows/blob/v3.5.5/workflow/controller/operator.go#L1209

---

> perhaps we leave one in and then refactor in a separate PR to avoid globals?

Sounds good to me

Reply:
Let's make sure to add a `// TODO:` comment here to modify this in the future. Can reference this thread in the comment: https://github.com/argoproj/argo-workflows/pull/12756#discussion_r1530245007.
This would be a really easy follow-up to miss (and may not  happen soon), so being explicit is good (and `TODO`s can be searched for in the codebase easily)